structuralcodes/sections/_section_integrators.py

```diff
@@ -0,0 +1,442 @@
+            strain (ArrayLike): The scalar strain components necessary for
+                describing the assumed strain distribution over the geometry.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def integrate(self, *prepared_input, **kwargs) -> float:
+        """Integrate stresses over the gemetry.
+
+        The input to this method is generated by the prepare_input method. It
+        also takes kwargs depending on the concrete implementation.
+
+        Arguments:
+            *prepared_input: The input prepared by the prepare_input method.
+
+        Keyword Arguments:
+            **kwargs: Keyword arguments depending on the concrete
+                implementation.
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def integrate_strain_response_on_geometry(
+        self, geo: CompoundGeometry, strain: ArrayLike
+    ):
+        """Integrate over the geometry to obtain the response due
+        to strains.
+        """
+        raise NotImplementedError
+
+
+class MarinIntegrator(SectionIntegrator):
+    """Section integrator based on the Marin algorithm."""
+
+    def prepare_input(
+        self, geo: CompoundGeometry, strain: ArrayLike
+    ) -> t.Tuple[t.Tuple[np.ndarray, np.ndarray, np.ndarray]]:
+        """Prepare general input to the integration.
+
+        Calculate the stresses based on strains in a set of points.
+
+        Args:
+            geo (CompoundGeometry): The geometry of the section.
+            strain (ArrayLike): The strains and curvatures of the section.
+                                given in the format (ea, ky, kz) which are
+                                strain at 0,0
+                                curvature y axis
+                                curvature z axis
+        """
+        # This method should do the following tasks:
+        # - For each geo:
+        #   - ask constitutive law strain limits and coefficients
+        #   - For each material part, the part should furthermore be split
+        # according to constant, linear and parabolic stress distribution.
+        #   - For each part collect coordinates y and z in separate np.ndarray
+        # iterables, and stress coefficients in a two-dimensional np.ndarray.
+        #
+        # The method should therefore return a tuple that collects the y, z,
+        # and stress coefficients for each part.
+        prepared_input = []
+        # 1. Rotate section in order to have neutral axis horizontal
+        angle = atan2(strain[2], strain[1])
+
+        rotated_geom = geo.rotate(angle)
+        # 2. Get y coordinate of neutral axis in this new CRS
+        strain_rotated = [strain[0], (strain[2] ** 2 + strain[1] ** 2) ** 0.5]
+
+        # 3. For each SurfaceGeometry on the CompoundGeometry:
+        for g in rotated_geom.geometries:
+            # 3a. get coefficients and strain limits from constitutive law
+            strains, coeffs = g.material.__marin__(strain_rotated)
```

Comment:
Here we should catch the `AttributeError` that is raised if the `__marin__` method is not defined.